cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseTable.java

```diff
@@ -317,37 +323,67 @@ private boolean hbaseFlush(List<Mutation> mutations)
     return false;
   }
 
-  private PutBuilder getPut(PutBuilder existing, byte[] row) {
+  private PutBuilder getPut(PutBuilder existing, byte[] row, @Nullable byte[] txId) {
     if (existing != null) {
       return existing;
     }
-    return tableUtil.buildPut(row)
-      .setAttribute(TX_MAX_LIFETIME_MILLIS_KEY, txMaxLifetimeMillis);
+    PutBuilder putBuilder = tableUtil.buildPut(row);
+    if (txId != null) {
+      putBuilder.setAttribute(TX_MAX_LIFETIME_MILLIS_KEY, txMaxLifetimeMillis)
+        .setAttribute(TX_ID, txId);
+    }
+     return putBuilder;
   }
 
   private PutBuilder getPutForIncrement(PutBuilder existing, byte[] row) {
     if (existing != null) {
       return existing;
     }
     return tableUtil.buildPut(row)
-      .setAttribute(DELTA_WRITE, Bytes.toBytes(true))
-      .setAttribute(TX_MAX_LIFETIME_MILLIS_KEY, txMaxLifetimeMillis);
+      .setAttribute(DELTA_WRITE, Bytes.toBytes(true));
   }
 
-  private IncrementBuilder getIncrement(IncrementBuilder existing, byte[] row, boolean transactional)
+  private IncrementBuilder getIncrement(IncrementBuilder existing, byte[] row, @Nullable byte[] txId,
+                                        @Nullable byte[] txWritePointer)
     throws IOException {
     if (existing != null) {
       return existing;
     }
     IncrementBuilder builder = tableUtil.buildIncrement(row).setAttribute(DELTA_WRITE, Bytes.toBytes(true));
-    builder.setAttribute(TX_MAX_LIFETIME_MILLIS_KEY, txMaxLifetimeMillis);
-    if (transactional) {
-      builder.setAttribute(WRITE_POINTER, Bytes.toBytes(tx.getWritePointer()));
+    if (txId != null) {
+      builder.setAttribute(TX_MAX_LIFETIME_MILLIS_KEY, txMaxLifetimeMillis);
+      builder.setAttribute(WRITE_POINTER, txWritePointer);
+      builder.setAttribute(TX_ID, txId);
       builder.setAttribute(TxConstants.TX_OPERATION_ATTRIBUTE_KEY, getEncodedTx());
     }
     return builder;
   }
 
+  /**
+   * Returns the transaction id for an operation.
+   *
+   * @return transaction id or -1 if not a transactional operation
+   */
+  public static long getTransactionId(OperationWithAttributes op, @Nullable Transaction tx) {
+    // CDAP does not encode transaction into a Put, a Delete or an Increment
+    // Instead sets the write pointer of the transaction as HBaseTable.WRITE_POINTER attribute
+    long txId;
+    if (tx == null) {
+      byte[] writePointerBytes = op.getAttribute(HBaseTable.TX_ID);
```

Comment:
you are getting the write pointer from the TX_ID field? That is in contradiction with the comment above.

Reply:
Right, need to fix the comment. It got outdated during a refactoring.